x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/adapters/fail_on_non_existing_function_call.ts

```diff
@@ -5,28 +5,24 @@
  * 2.0.
  */
 
-import { noop } from 'lodash';
-import { forkJoin, last, Observable, shareReplay, tap } from 'rxjs';
-import {
-  ChatCompletionChunkEvent,
-  createFunctionNotFoundError,
-  FunctionDefinition,
-} from '../../../../common';
-import { TokenCountEvent } from '../../../../common/conversation_complete';
+import { ignoreElements, last, merge, Observable, shareReplay, tap } from 'rxjs';
+import { createFunctionNotFoundError, FunctionDefinition } from '../../../../common';
+import { ChatEvent } from '../../../../common/conversation_complete';
 import { concatenateChatCompletionChunks } from '../../../../common/utils/concatenate_chat_completion_chunks';
-import { rejectTokenCountEvents } from '../../util/reject_token_count_events';
+import { withoutTokenCountEvents } from '../../../../common/utils/without_token_count_events';
 
 export function failOnNonExistingFunctionCall({
   functions,
 }: {
   functions?: Array<Pick<FunctionDefinition, 'name' | 'description' | 'parameters'>>;
 }) {
-  return (source$: Observable<ChatCompletionChunkEvent | TokenCountEvent>) => {
-    return new Observable<ChatCompletionChunkEvent | TokenCountEvent>((subscriber) => {
-      const shared = source$.pipe(shareReplay());
+  return (source$: Observable<ChatEvent>) => {
+    const shared$ = source$.pipe(shareReplay());
```

Comment:
Q: is the fact that we use shareReplay everywhere cause for concern? This holds on to every emission for the lifetime of the request so can increase memory usage quite significantly. And if we don't properly end the observable we could end up with a memory leak.

Reply:
Good question and something we should consider. In this case that's fine I think, as the Observable is short-lived, it doesn't keep on emitting values indefinitely.